sycl/include/sycl/nd_item.hpp

```diff
@@ -38,6 +41,492 @@ namespace ext::oneapi::experimental {
+    return LinId;
+  }
+
+  group<Dimensions> get_group() const {
+    // TODO: ideally Group object should be stateless and have a contructor with
+    // no arguments.
+    return detail::Builder::createGroup(get_global_range(), get_local_range(),
+                                        get_group_range(), get_group_id());
+  }
+
+  sub_group get_sub_group() const { return sub_group(); }
+
+  size_t __SYCL_ALWAYS_INLINE get_group(int Dimension) const {
+    size_t Id = get_group_id()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_linear_id() const {
+    size_t LinId = 0;
+    id<Dimensions> Index = get_group_id();
+    range<Dimensions> Extent = get_group_range();
+    if (1 == Dimensions) {
+      LinId = Index[0];
+    } else if (2 == Dimensions) {
+      LinId = Index[0] * Extent[1] + Index[1];
+    } else {
+      LinId =
+          Index[0] * Extent[1] * Extent[2] + Index[1] * Extent[2] + Index[2];
+    }
+    __SYCL_ASSUME_INT(LinId);
+    return LinId;
+  }
+
+  range<Dimensions> get_group_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initNumWorkgroups<Dimensions, range<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_range(int Dimension) const {
+    size_t Range = get_group_range()[Dimension];
+    __SYCL_ASSUME_INT(Range);
+    return Range;
+  }
+
+  range<Dimensions> get_global_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalSize<Dimensions, range<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  size_t get_global_range(int Dimension) const {
+    size_t Val = get_global_range()[Dimension];
+    __SYCL_ASSUME_INT(Val);
+    return Val;
+  }
+
+  range<Dimensions> get_local_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initWorkgroupSize<Dimensions, range<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  size_t get_local_range(int Dimension) const {
+    size_t Id = get_local_range()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
+  id<Dimensions> get_offset() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalOffset<Dimensions, id<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  nd_range<Dimensions> get_nd_range() const { return nd_range<Dimensions>(); }
+
+  void barrier(access::fence_space accessSpace =
+                   access::fence_space::global_and_local) const {
+    uint32_t flags = _V1::detail::getSPIRVMemorySemanticsMask(accessSpace);
+    __spirv_ControlBarrier(__spv::Scope::Workgroup, __spv::Scope::Workgroup,
+                           flags);
+  }
```

Comment:
Should this include some sort of deprecation notice?  SYCL 2020 removed this function completely, in favor of `sycl::group_barrier`.

Reply:
If so, we need to change stateful implementation too. As it's not directly related to making `nd_item` stateless, I created https://github.com/intel/llvm/issues/12531 to make this change separately.
@intel/sycl-language-enabling-triage, FYI.

Reply:
Yes, we have an internal task/bug report for the deprecation and removal of the method. We have tried it once already (even just deprecation), but reverted the change under pressure from some customers. The problem with `group_barrier` is that it doesn't accept fence space argument and uses the widest scope possible for memory fence. This causes performance regressions on some of our HW.

We should have a bug reported against corresponding low-level runtimes, but I'm not sure if there has been any meaningful progress on it.

Reply:
Those are two different kind of scopes/spaces:

[`OpControlBarrier`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpControlBarrier) accepts 3 arguments:
1. execution scope - which work-items should wait for other work-items
2. memory scope - how "wide" mem-fence should be: do we care about synchronizing memory operations what may have happened in other work-groups, or just this one is fine?
3. memory semantics - address spaces to synchronize + memory order (acquire, release, etc.)

[`access::fence_space`](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/access/access.hpp#L43) is essentially an address space, which needs to be synchronized and it is a legacy from SYCL 1.2.1. As you can see on the snippet above, `nd_item::barrier` does execution synchronization of a work-group and mem-fence within a work-group for *the specified address space*.

[`group_barrier`](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_group_barrier) accepts `memory_scope`, which is how wide the mem fence should be, but not which address spaces it should cover.

With `group_barrier` we end up doing execution synchronization of a work-group and mem-fence within a specified scope for *all address spaces*. See [1](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/group_barrier.hpp), [2](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/spirv.hpp#L1016)

`group_barrier` leads to emission of some extra mem fences and if I understand correctly, there is a bug/missing optimization in IGC that does not remove those extra fences, which leads to extra synchronization and performance degradation. Since `group_barrier` does not allow to pass an address space, users have no workaround (except for using the legacy `nd_item::barrier`) for that bug.

Reply:
@AlexeySachkov, are you saying that to fix that bug we need to extend `group_barrier` API? Could you open a separate issue for that, please?

Reply:
> @AlexeySachkov, are you saying that to fix that bug we need to extend `group_barrier` API? Could you open a separate issue for that, please?

SYCL 2020 version of the barrier definitely has less capabilities than SYCL 1.2.1 `nd_item::barrier` and that puts an extra burden on the implementation to maintain the same level of performance. However, as I understand it, that was a deliberate choice in the spec direction to have a single happens-before relationship for all address spaces at once to simplify user experience with SYCL. It comes with a cost, unless mitigated by optimizations made at different levels of SW stack.

@Pennycook was the one who led that effort,  I believe. See internal Khronos PR 520 (removes mentions of address spaces from `group_barrier`), issue 354 (prompted PR 520) and PR 420 (first (?) mention of single happens-before relationship for all address spaces).

Reply:
Yes, this was deliberate.  There were lots of reasons, but my favorite example is that the combination of C++ templates, generic address spaces and a split happens-before makes it _really_ easy to write code with hard-to-find bugs in it:
```c++
template <typename T1, typename T2>
void buggy(nd_item<1> it, T1* ptr1, T2* ptr2) {
  if (it.get_global_id(0)) ptr1 = x;
  if (it.get_global_id(0)) ptr2 = y;
  nd_item::barrier(access::global_space); // oops
  if (ptr1 == x && ptr2 == y) {
    /* do something */
  }
}
```

This code will work if both pointers happen to point to the global address space, but may fail if either pointer points to something else.  Detecting which space to fence statically would require the user to opt-in to using address space qualifiers (which they typically don't want to do).  Detecting which space to fence dynamically would require runtime checks of certain pointers.  Calling `nd_item::barrier()` to fence both spaces is the easy fix, and that's equivalent to calling `group_barrier`.

There may be some special corner cases where somebody who knows what they're doing has a reason to generate a fence for one address space.  But my view is that if a developer is capable of fully understanding the ramifications of a split-happens before on a specific architecture, they should also be capable of calling something like inline assembly or SPIR-V intrinsic.